core/src/main/java/org/apache/iceberg/UpdateSnapshotReferencesOperation.java

```diff
@@ -0,0 +1,166 @@
+        .exponentialBackoff(
+            base.propertyAsInt(COMMIT_MIN_RETRY_WAIT_MS, COMMIT_MIN_RETRY_WAIT_MS_DEFAULT),
+            base.propertyAsInt(COMMIT_MAX_RETRY_WAIT_MS, COMMIT_MAX_RETRY_WAIT_MS_DEFAULT),
+            base.propertyAsInt(COMMIT_TOTAL_RETRY_TIME_MS, COMMIT_TOTAL_RETRY_TIME_MS_DEFAULT),
+            2.0 /* exponential */)
+        .onlyRetryOn(CommitFailedException.class)
+        .run(taskOps -> {
+          Map<String, SnapshotRef> refs = apply();
+          TableMetadata.Builder updatedMetadata = TableMetadata.buildFrom(base);
+          updatedMetadata.
+          taskOps.commit(base, updated);
+        });
+  }
+
+  public UpdateSnapshotReferencesOperation createBranch(String name, long snapshotId) {
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    SnapshotRef branch = SnapshotRef.branchBuilder(snapshotId).build();
+    refsToUpdate.put(name, branch);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation createTag(String name, long snapshotId) {
+    Preconditions.checkNotNull(name, "Tag name cannot be null");
+    SnapshotRef tag = SnapshotRef.tagBuilder(snapshotId).build();
+    refsToUpdate.put(name, tag);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation removeBranch(String name) {
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    SnapshotRef ref = base.refs().get(name);
+    refsToRemove.add(name);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation removeTag(String name) {
+    Preconditions.checkNotNull(name, "Tag name cannot be null");
+    SnapshotRef ref = base.refs().get(name);
+    refsToRemove.add(name);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation renameBranch(String name, String newName) {
+    Preconditions.checkNotNull(name, "Tag name cannot be null");
+    Preconditions.checkNotNull(newName, "Tag name cannot be null");
+    SnapshotRef ref = base.refs().get(name);
+    refsToUpdate.put(newName, ref);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation replaceBranch(String branchName, long snapshotId) {
+    Preconditions.checkNotNull(branchName, "Branch name cannot be null");
+    Preconditions.checkNotNull(newName, "Tag name cannot be null");
+    SnapshotRef ref = base.refs().get(name);
+    refsToUpdate.put(newName, ref);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation replaceTag(String tagName, long snapshotId) {
+    Preconditions.checkNotNull(tagName, "Tag name cannot be null");
+    SnapshotRef ref = base.refs().get(name);
+    refsToUpdate.put(newName, ref);
+    return this;
+  }
+
+  public UpdateSnapshotReferencesOperation setMinSnapshotsToKeep(String name, int minSnapshotsToKeep) {
+    Preconditions.checkNotNull(name, "Branch name cannot be null");
+    SnapshotRef ref = base.refs().get(name);
+    SnapshotRef updateBranch = SnapshotRef.builderFrom(ref)
+        .minSnapshotsToKeep(minSnapshotsToKeep)
```

Comment:
Does this throw a reasonable error message when the ref is a tag?

Reply:
https://github.com/amogh-jahagirdar/iceberg/blob/master/api/src/main/java/org/apache/iceberg/SnapshotRef.java#L156

It throws "Tags do not support setting minSnapshotsToKeep" which I think is reasonable to surface